app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt

```diff
@@ -368,20 +411,58 @@ class WallpapersUseCases(
          *
          * @param wallpaper The selected wallpaper.
          */
-        operator fun invoke(wallpaper: Wallpaper)
+        suspend operator fun invoke(wallpaper: Wallpaper)
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal class LegacySelectWallpaperUseCase(
+        private val settings: Settings,
+        private val store: AppStore,
+    ) : SelectWallpaperUseCase {
+        /**
+         * Select a new wallpaper. Storage and the store will be updated appropriately.
+         *
+         * @param wallpaper The selected wallpaper.
+         */
+        override suspend fun invoke(wallpaper: Wallpaper) {
+            settings.currentWallpaperName = wallpaper.name
+            store.dispatch(AppAction.WallpaperAction.UpdateCurrentWallpaper(wallpaper))
+            Wallpapers.wallpaperSelected.record(
+                Wallpapers.WallpaperSelectedExtra(
+                    name = wallpaper.name,
+                    themeCollection = wallpaper.collection.name
+                )
+            )
+        }
     }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal class DefaultSelectWallpaperUseCase(
         private val settings: Settings,
         private val store: AppStore,
+        private val fileManager: WallpaperFileManager,
+        private val downloader: WallpaperDownloader,
     ) : SelectWallpaperUseCase {
         /**
          * Select a new wallpaper. Storage and the store will be updated appropriately.
          *
          * @param wallpaper The selected wallpaper.
          */
-        override fun invoke(wallpaper: Wallpaper) {
+        override suspend fun invoke(wallpaper: Wallpaper) {
+            if (fileManager.wallpaperImagesExist(wallpaper)) {
```

Comment:
I would be willing to revisit this decision in the future, but I think it would be a rather large refactor to do at this stage. I also personally have some gripes with `interactor/controller` design patterns, as I think they can tend to add unnecessary abstraction layers just for the sake of them. We should talk about this more when I get back 🙂